controller/kubernetes_pod_controller.go

```diff
@@ -0,0 +1,273 @@
+	clientset "k8s.io/client-go/kubernetes"
+	v1core "k8s.io/client-go/kubernetes/typed/core/v1"
+	listerv1 "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/kubernetes/pkg/controller"
+
+	"github.com/longhorn/longhorn-manager/datastore"
+	"github.com/longhorn/longhorn-manager/types"
+)
+
+const (
+	controllerAgentName = "Longhorn Kubernetes Pod Controller"
+)
+
+type KubernetesPodController struct {
+	// use as the OwnerID of the controller
+	controllerID string
+
+	kubeClient    clientset.Interface
+	eventRecorder record.EventRecorder
+
+	ds *datastore.DataStore
+
+	pLister   listerv1.PodLister
+	pvLister  listerv1.PersistentVolumeLister
+	pvcLister listerv1.PersistentVolumeClaimLister
+
+	pStoreSynced   cache.InformerSynced
+	pvStoreSynced  cache.InformerSynced
+	pvcStoreSynced cache.InformerSynced
+
+	queue workqueue.RateLimitingInterface
+
+	//// for unit test
+	//nowHandler func() string
+}
+
+func NewKubernetesPodController(
+	ds *datastore.DataStore,
+	scheme *runtime.Scheme,
+	kubePodInformer coreinformers.PodInformer,
+	kubePersistentVolumeInformer coreinformers.PersistentVolumeInformer,
+	kubePersistentVolumeClaimInformer coreinformers.PersistentVolumeClaimInformer,
+	kubeClient clientset.Interface,
+	controllerID string) *KubernetesPodController {
+
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(logrus.Infof)
+	// TODO: remove the wrapper when every clients have moved to use the clientset.
+	eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{
+		Interface: v1core.New(kubeClient.CoreV1().RESTClient()).Events(""),
+	})
+
+	kc := &KubernetesPodController{
+		controllerID: controllerID,
+
+		ds: ds,
+
+		kubeClient:    kubeClient,
+		eventRecorder: eventBroadcaster.NewRecorder(scheme, v1.EventSource{Component: controllerAgentName}),
+
+		pLister:   kubePodInformer.Lister(),
+		pvLister:  kubePersistentVolumeInformer.Lister(),
+		pvcLister: kubePersistentVolumeClaimInformer.Lister(),
+
+		pStoreSynced:   kubePodInformer.Informer().HasSynced,
+		pvStoreSynced:  kubePersistentVolumeInformer.Informer().HasSynced,
+		pvcStoreSynced: kubePersistentVolumeClaimInformer.Informer().HasSynced,
+
+		queue: workqueue.NewNamedRateLimitingQueue(EnhancedDefaultControllerRateLimiter(), "longhorn-kubernetes-pod"),
+	}
+
+	kubePodInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			pod := obj.(*v1.Pod)
+			kc.enqueuePodChange(pod)
+		},
+		UpdateFunc: func(old, cur interface{}) {
+			curPod := cur.(*v1.Pod)
+			kc.enqueuePodChange(curPod)
+		},
+		DeleteFunc: func(obj interface{}) {
+			pod := obj.(*v1.Pod)
+			kc.enqueuePodChange(pod)
+		},
+	})
+
+	return kc
+}
+
+func (kc *KubernetesPodController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer kc.queue.ShutDown()
+
+	logrus.Infof("Start " + controllerAgentName)
```

Comment:
log format

Reply:
Fixed

Reply:
sorry the intent was not to remove the usage of the const, but instead since you are calling Infof which expects a format string, to just use the "Start %s", controllerAgentName instead of the unecessary string concatenation.